components/tools/OmeroWeb/test/integration/test_api_containers.py

```diff
@@ -0,0 +1,113 @@
+        encoder = get_encoder(obj.__class__)
+        expected.append(encoder.encode(obj))
+    return expected
+
+
+def assert_objects(conn, json_objects, omero_ids_objects, dtype="Project",
+                   group='-1'):
+    """
+    Load objects from OMERO, via conn.getObjects().
+
+    marshal with omero_marshal and compare with json_objects.
+    omero_ids_objects can be IDs or list of omero.model objects.
+    """
+    pids = []
+    for p in omero_ids_objects:
+        try:
+            pids.append(long(p))
+        except TypeError:
+            pids.append(p.id.val)
+    conn.SERVICE_OPTS.setOmeroGroup(group)
+    projects = conn.getObjects(dtype, pids, respect_order=True)
+    projects = [p._obj for p in projects]
+    expected = marshal_objects(projects)
+    assert len(json_objects) == len(expected)
+    for o1, o2 in zip(json_objects, expected):
+        assert o1 == o2
+
+
+class TestContainers(IWebTest):
+    """Tests querying & editing Datasets, Screens etc."""
+
+    @pytest.mark.parametrize("dtype", ['Project', 'Dataset', 'Screen'])
+    def test_container_crud(self, dtype):
+        """
+        Test create, read, update and delete of Containers.
+
+        Create with POST to /save
+        Read with GET of /m/dtype/:id/
+        Update with PUT to /m/dtype/:id/
+        Delete with DELETE to /m/dtype/:id/
+        """
+        django_client = self.django_root_client
+        group = self.ctx.groupId
+        version = settings.API_VERSIONS[-1]
+        # Need to get the Schema url to create @type
+        base_url = reverse('api_base', kwargs={'api_version': version})
+        rsp = _get_response_json(django_client, base_url, {})
+        schema_url = rsp['schema_url']
+        # specify group via query params
+        save_url = "%s?group=%s" % (rsp['save_url'], group)
+        project_name = 'test_container_create_read'
+        payload = {'Name': project_name,
+                   '@type': '%s#%s' % (schema_url, dtype)}
+        rsp = _csrf_post_json(django_client, save_url, payload,
+                              status_code=201)
+        # We get the complete new Object returned
+        assert rsp['Name'] == project_name
+        object_id = rsp['@id']
+
+        # Read Object
+        object_url = "%sm/%ss/%s/" % (base_url, dtype.lower(), object_id)
+        print object_url
+        object_json = _get_response_json(django_client, object_url, {})
+        assert object_json['@id'] == object_id
+        conn = BlitzGateway(client_obj=self.root)
+        assert_objects(conn, [object_json], [object_id], dtype=dtype)
+
+        # Update Object...
+        object_json['Name'] = 'new name'
+        rsp = _csrf_put_json(django_client, save_url, object_json)
+        # ...and read again to check
+        updated_json = _get_response_json(django_client, object_url, {})
+        assert updated_json['Name'] == 'new name'
+
+        # Delete
+        _csrf_delete_response_json(django_client, object_url, {})
```

Comment:
While we are testing 'unsafe' HTTP operations I think we should be testing both cases especially when csrf token is not passed. If you feel is not convenient to write multiple lines of code every time how about if we add wrapper:

```
delete_response_json(...):
    _csrf_delete_response_json(...)
    _delete_response_json(...)
```